### PR TITLE
New SeekableSource class for efficient file reading.

### PR DESCRIPTION
**Pull request by jkinkead**
On my laptop, this outperforms `scala.io.Source` reading character-wise by a factor of 3.4, and line-wise by a factor of 1.7.

If you want to see the slightly out-of-date version with a memory map option, you can check out [this commit](https://github.com/jkinkead/common/tree/2006bb1c2ee0851ef9c978a52347ff43a670cab5). That's a little slower for the line iterator, because it's not using the dirty tricks I gleaned from BufferedReader. It's still valid for comparing the memory map option with the standard option, though.

### CODE REVIEW

File: core/src/main/scala/org/allenai/common/SeekableSource.scala [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+  * <p>This also has a `getLines()` method to iterate over lines. Any reads done on this iterator
+  * are reflected in the main Source, and reads on the Source are reflected in the
+  * `Iterator[String]`.</p>
+  * @param inFile the file channel to wrap
+  * @param bufferSize the size of the internal buffer to use
+  * @param codec the codec to use. Must be one of UTF-8 or ISO-8859-1.
+  * @throws IllegalArgumentException if `bufferSize` is less than 3, or `codec` is not UTF-8 or
+  * ISO-8859-1
+  */
+class SeekableSource(inFile: FileChannel, bufferSize: Int = 8192)(implicit codec: Codec)
```
Comment (dirkgr) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: Couldn't you have a buffer much larger than this? Why not 8MB?

File: core/src/main/scala/org/allenai/common/SeekableSource.scala [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    // Ensure we are either at the end of file, or have at least three bytes that are readable.
+    if (inBuffer.remaining() < 3 && inputRemaining) {
+      fillBuffer()
+    }
+    // See https://en.wikipedia.org/wiki/UTF-8 for details. This is checking the first byte for the
+    // total length of the encoding.
+    val first = inBuffer.get()
+    val intVal = if ((first & 0x80) == 0) {
+      return first.toChar
+    } else if ((first & 0xe0) == 0xc0) {
```
Comment (dirkgr) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: This is conditional of the byte starting with `110`. Is that what you want? It's not what the comment says below.

File: core/src/main/scala/org/allenai/common/SeekableSource.scala [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+      // TODO(jkinkead): Secondary bytes (here and below) should all have their two highest bits set
+      // as 10. This should be asserted somehow, and if they're corrupted, producing an unknown
+      // character instead.
+      val second = inBuffer.get()
+      ((first & 0x1f) << 6) | (second & 0x3f)
+    } else if ((first & 0xf0) == 0xe0) {
```
Comment (dirkgr) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: And this starts with `1110`. Maybe say that somewhere?

File: core/src/main/scala/org/allenai/common/SeekableSource.scala [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+      // won't fit in a JVM Char.
+
+      // skipCount stores the *additional* bytes we will skip, not the total length of the encoding.
+      val skipCount = if ((first & 0xf8) == 0xf0) {
+        3
+      } else if ((first & 0xfc) == 0xf8) {
+        4
+      } else if ((first & 0xfe) == 0xfc) {
+        5
+      } else {
```
Comment (dirkgr) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: I read somewhere that UTF-8 was designed so you can pick up a stream in the middle. There is some way to figure out where the next character starts. If that's true, you shouldn't have to do this. On the other hand, we're in the realm of illegal UTF-8, so maybe that doesn't apply here anyways.

File: core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import java.nio.file.{ Files, StandardOpenOption }
+
+import scala.io.Codec
+
+class SeekableSourceSpec extends UnitSpec {
+  // UTF-8 is the default encoding for these tests.
+  implicit val codec = Codec.UTF8
+
+  /** Stores füü, in UTF-8. */
+  val foomlaut: Array[Byte] = Array('f'.toByte, 0xc3, 0xbc, 0xc3, 0xbc) map { _.toByte }
```
Comment (dirkgr) [inline core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala]: What about characters that span three bytes?

File: core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    val windowsNewlines = newFileWithChars("foo\rbar\r\ngaz\r")
+    val source = new SeekableSource(windowsNewlines)
+    val lines = source.getLines()
+
+    lines.next() should be("foo")
+    lines.next() should be("bar")
+    lines.next() should be("gaz")
+    lines.hasNext should be(false)
+  }
+}
```
Comment (dirkgr) [inline core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala]: What about a test for the malformed UTF-8 case?

**Comment by dirkgr** [comment]
I'm assuming you smoke-tested this on the Wikipedia corpus, to make sure it doesn't blow up?

**Comment by jkinkead** [comment]
Yeah, my test numbers come from reading the wikipedia corpus - and verifying line counts with `scala.io.Source`.

File: core/src/main/scala/org/allenai/common/SeekableSource.scala (hunk 1 above)
Comment (jkinkead) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: I think I'm just living in the past . . . :P

This size is what `java.io.BufferedReader` uses. Otherwise, it's pretty arbitrary. I wasn't seeing huge boosts when I was playing around with sizes, but 8M seems like a totally sensible default.

File: core/src/main/scala/org/allenai/common/SeekableSource.scala (hunk 2 above)
Comment (jkinkead) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: The comment is talking about secondary bytes (all bytes following the first size-indicating byte). I'll try to clarify (and address the TODO).

File: core/src/main/scala/org/allenai/common/SeekableSource.scala (hunk 3 above)
Comment (jkinkead) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: Done.

File: core/src/main/scala/org/allenai/common/SeekableSource.scala (hunk 4 above)
Comment (jkinkead) [inline core/src/main/scala/org/allenai/common/SeekableSource.scala]: Actually, yes, you can skip any bytes that start with the secondary-character sigil. I'll do that.

File: core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala (hunk 5 above)
Comment (jkinkead) [inline core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala]: Done.

File: core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala (hunk 6 above)
Comment (jkinkead) [inline core/src/test/scala/org/allenai/common/SeekableSourceSpec.scala]: Done.

**Comment by jkinkead** [comment]
OK, comments addressed, and unit tests expanded! I also fixed a couple of bugs.

It turns out that Java *does* support all characters in UTF-16, so I updated the by-character reader to accommodate this. Did you want to take another look?

**Comment by dirkgr** [comment]
UTF is messy business.

**Comment by dirkgr** [comment]
When you have all the special cases handled, make sure it's still faster than what you're replacing :-)

**Comment by jkinkead** [comment]
What's really strange is that the comments you made on the specific commits went to my gmail account . . . :)

I'll let you know when I've added tests for the other corner-cases (and fixed them if need be).

**Comment by jkinkead** [comment]
Special-case tested (and handled).

**Comment by dirkgr** [comment]
At this point, I have no further ideas about how to break this. LGTM!

**Comment by jkinkead** [comment]
Sweet!
